common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -0,0 +1,171 @@
+    public final String version;
+
+    /**
+     * Collation sensitive hash function. Output for two UTF8Strings will be the same if they are
+     * equal according to the collation.
+     */
+    public final Function<UTF8String, Integer> hashFunction;
+
+    /**
+     * Potentially faster way than using comparator to compare two UTF8Strings for equality.
+     * Falls back to binary comparison if the collation is binary.
+     */
+    public final BiFunction<UTF8String, UTF8String, Boolean> equalsFunction;
+
+    /**
+     * Binary collation implies that UTF8Strings are considered equal only if they are
+     * byte for byte equal. All accent or case-insensitive collations are considered non-binary.
+     */
+    public final boolean isBinaryCollation;
+
+    public CollationInfo(
+      String collationName,
+      Collator collator,
+      Comparator<UTF8String> comparator,
+      String version,
+      Function<UTF8String, Integer> hashFunction,
+      boolean isBinaryCollation) {
+      this.collationName = collationName;
+      this.collator = collator;
+      this.comparator = comparator;
+      this.version = version;
+      this.hashFunction = hashFunction;
+      this.isBinaryCollation = isBinaryCollation;
+
+      if (isBinaryCollation) {
+        this.equalsFunction = UTF8String::binaryEquals;
+      } else {
+        this.equalsFunction = (s1, s2) -> this.comparator.compare(s1, s2) == 0;
+      }
+    }
+
+    /**
+     * Constructor with comparators that are inherited from the given collator.
+     */
+    public CollationInfo(
+      String collationName, Collator collator, String version, boolean isBinaryCollation) {
+      this(
+        collationName,
+        collator,
+        (s1, s2) -> collator.compare(s1.toString(), s2.toString()),
+        version,
+        s -> collator.getCollationKey(s.toString()).hashCode(),
+        isBinaryCollation);
+    }
+  }
+
+  private final CollationInfo[] collatorTable;
+  private final Hashtable<String, Integer> collationNameToIdMap = new Hashtable<>();
+
+  private CollationFactory() {
+    collatorTable = new CollationInfo[4];
+
+    // Binary comparison. This is the default collation.
+    // No custom comparators will be used for this collation.
+    // Instead, we rely on byte for byte comparison.
+    collatorTable[0] = new CollationInfo(
+      "UCS_BASIC",
+      null,
+      UTF8String::binaryCompare,
+      "1.0",
+      UTF8String::binaryHash,
+      true);
+
+    // Case-insensitive UTF8 binary collation.
+    // TODO: Do in place comparisons instead of creating new strings.
+    collatorTable[1] = new CollationInfo(
+      "UCS_BASIC_LCASE",
+      null,
+      (s1, s2) -> s1.toLowerCase().binaryCompare(s2.toLowerCase()), "1.0",
+      (s) -> s.toLowerCase().binaryHash(),
+      false);
+
+    // UNICODE case sensitive comparison (ROOT locale, in ICU).
+    collatorTable[2] = new CollationInfo(
+      "UNICODE", Collator.getInstance(ULocale.ROOT), "153.120.0.0", true);
```

Comment:
how do we get the version number?

Reply:
I mean, how do you get `"153.120.0.0"`?

Reply:
It represents a hard coded version of collator used for this collation. e.g. `Collator.getInstance(ULocale.GERMAN).getVersion` will return "153.120.0.0" with current version of ICU. I added additional comment in the `version` field header.